services/brig/test/integration/API/User/Auth.hs

```diff
@@ -227,7 +230,51 @@ testThrottleLogins conf b = do
     liftIO $ do
         assertBool "throttle delay" (n > 0)
         threadDelay (1000000 * (n + 1))
-    void $ login b (defEmailLogin e) SessionCookie
+    login b (defEmailLogin e) SessionCookie !!! const 200 === statusCode
+
+testLimitRetries :: HasCallStack => Maybe Opts.Opts -> Brig -> Http ()
+testLimitRetries (Just conf) brig = do
+    let Just opts = Opts.setLimitFailedLogins . Opts.optSettings $ conf
+    unless (Opts.timeout opts <= 30) $
+        error "`loginRetryTimeout` is the number of seconds this test is running.  Please pick a value < 30."
+
+    usr <- randomUser brig
+    let Just email = userEmail usr
+
+    usr' <- randomUser brig
+    let Just email' = userEmail usr'
+
+    -- Login 5 times with bad password.
+    forM_ [1..Opts.retryLimit opts] $ \_ ->
+        login brig (emailLogin email defWrongPassword (Just defCookieLabel)) SessionCookie
+            <!! const 403 === statusCode
+
+    -- Login once more. This should fail for usr, even though password is correct...
+    resp <- login brig (defEmailLogin email) SessionCookie
+        <!! const 403 === statusCode
+    -- ...  but not for usr'!
+    login brig (defEmailLogin email') SessionCookie
+        !!! const 200 === statusCode
+
+    -- After the amount of time specified in "Retry-After", though,
+    -- throttling should stop and login should work again
+    do  let Just retryAfterSecs = fromByteString =<< getHeader "Retry-After" resp
+            retryTimeout = Opts.Timeout $ fromIntegral retryAfterSecs
+        liftIO $ do
+            assertEqual "throttle delay (1)" retryTimeout (Opts.timeout opts)
+            threadDelay (1000000 * (retryAfterSecs - 1))
+
+    -- fail again later into the block time window
+    rsp <- login brig (defEmailLogin email) SessionCookie <!! const 403 === statusCode
+    do  let Just retryAfterSecs = fromByteString =<< getHeader "Retry-After" rsp
```

Comment:
In principle, `retryAfterSecs` has a value of ~1 here, right? Won't `retryAfterSecs - 2` risk being negative? What will happen with `threadDelay` if given a negative value? My empirical tests show that `threadDelay` with a negative value == no-op so I guess this is fine given that you assert in the line above.

Reply:
does https://github.com/wireapp/wire-server/pull/830/commits/eec445408881a39bef043c089f91eb2afcca2d7d answer your question?

(it's also a lot more correct, so thanks for the question.  :))

Reply:
does https://github.com/wireapp/wire-server/pull/830/commits/eec445408881a39bef043c089f91eb2afcca2d7d answer your question?

(it's also a lot more correct, so thanks for the question.  :))